### ISSUE TITLE
Unprepare execute on pool

### ISSUE THREAD

**Issue by shelru** [sidorares/node-mysql2#959]
```js
const info = await db.promisePool.execute(query2,updateArr);
        if(info[0].affectedRows){
            db.promisePool.unprepare(query2);
```

`TypeError: _dbcontroller.default.promisePool.unprepare is not a function`

**Comment by shelru** [comment]
@sidorares

**Comment by sidorares** [comment]
You'll have to do this:

```js
const conn = await db.promisePool.getConnection(); 
const info = await conn.execute(query2,updateArr);
if(info[0].affectedRows) {
    conn.unprepare(query2);
}
```

Is there any reason you want to clear prepared statement manually?

**Comment by shelru** [comment]
Yes, too much dynamic select's with different order.

**Comment by shelru** [comment]
@sidorares 
```js

**Comment by shelru** [comment]
So, i get, only query and format can work with `??`, can you write 'bout it in readme?

**Comment by shelru** [comment]
```js
SELECT `id`, ?, ?, ?, ?, ?, ?, ?, ? FROM `users` WHERE `id` = ?
[ '`name`',
  '`surename`',
  '`type`',
  '`type2`',
  '`fname`',
  '`group`',
  '`active`',
  '`login`',
  1 ]
[ [ BinaryRow { id: 1, '?': '`login`' } ],
  [ { catalog: 'def',
      schema: 'tests',
      name: 'id',
      orgName: 'id',
      table: 'users',
      orgTable: 'users',
      characterSet: 63,
      columnLength: 10,
      columnType: 3,
      flags: 16935,
      decimals: 0 },
    { catalog: 'def',
      schema: '',
      name: '?',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 224,
      columnLength: 24,
      columnType: 253,
      flags: 1,
      decimals: 31 },
    { catalog: 'def',
      schema: '',
      name: '?',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 224,
      columnLength: 40,
      columnType: 253,
      flags: 1,
      decimals: 31 },
    { catalog: 'def',
      schema: '',
      name: '?',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 224,
      columnLength: 24,
      columnType: 253,
      flags: 1,
      decimals: 31 },
    { catalog: 'def',
      schema: '',
      name: '?',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 224,
      columnLength: 28,
      columnType: 253,
      flags: 1,
      decimals: 31 },
    { catalog: 'def',
      schema: '',
      name: '?',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 224,
      columnLength: 28,
      columnType: 253,
      flags: 1,
      decimals: 31 },
    { catalog: 'def',
      schema: '',
      name: '?',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 224,
      columnLength: 28,
      columnType: 253,
      flags: 1,
      decimals: 31 },
    { catalog: 'def',
      schema: '',
      name: '?',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 224,
      columnLength: 32,
      columnType: 253,
      flags: 1,
      decimals: 31 },
    { catalog: 'def',
      schema: '',
      name: '?',
      orgName: '',
      table: '',
      orgTable: '',
      characterSet: 224,
      columnLength: 28,
      columnType: 253,
      flags: 1,
      decimals: 31 } ] ]
```
returned, why?
Is there anyway to make column's dynamic too?
```js

        Object.entries(req.body.columns).forEach(([key,value]) => {
                if(key.startsWith('p'))return;
                if(!value)return;
                query = `${query}, ?`;
                updateArr.push(`\`${key}\``);
            }
        );
        query = `${query} FROM \`users\` WHERE \`id\` = ?`;
        console.log(query);
        updateArr.push(req.body.id);
        console.log(updateArr);
        
        const conn = await db.promisePool.getConnection(); 
        const rows = await conn.execute(query,updateArr);
        console.log(rows);
```

**Comment by afwn90cj93201nixr2e1re** [comment]
No, i think no reasons for that, but im still dont get how LRU work and where is placed(you said that on mysql server, but seems like that's not true(?)).

So, there dynamic queries that's the main reason why i unpreparing it's everytime.


> https://github.com/sidorares/node-mysql2/commit/dd8dc361d7ebe614f332c0dbaf1de22cad8f86f3

So, that's shown that prepared statements can cause error's throwing?

**Comment by thibauds** [comment]
@damianobarbati it seems valid to me, you should be able to do:

```js
const pool = mysql.createPool(options)
const conn = await pool.getConnection()
const result = await conn.execute(query, data)
conn.unprepare(query)
```



@sidorares do you have any info about the LRU cache because we run into an issue in production: 
- we have `n` instances of our app using the `mysql2` package
- each instance is using a pool with `c` connections
- prepared statements are per connection

Thus we have `n` x `c` x `distinct(queries)` prepared statements which can go above our max_prepared_statement_count.

I was able to reproduce the issue with:
```js
  const pool = mysql.createPool(options)
  let ids = []
  for (let i = 0; i < maxPreparedStatement; i++) {
    ids.push(i)
    let query = `SELECT * FROM Dummy WHERE id IN (${Array(ids.length).fill('?').join(',')})`
    const conn = await pool.getConnection()
    await conn.execute(query, ids)
    conn.release()
  }
 ```
A temporary solution could be to manually unprepare some statement but I would like to use a smarter way like an LRU cache! (Another solution could be to use proxysql).

**Comment by sidorares** [comment]
@thibauds whenever insert results in the statements cache exceeding `maxPreparedStatements` the oldest accessed statement is unprepared -

https://github.com/sidorares/node-mysql2/blob/ab3feb75bb4af1f6d288c0494a99aa2483ac24d5/lib/connection.js#L74

Maybe solution for you is to just lower `maxPreparedStatements` value? [The default is 16000](https://github.com/sidorares/node-mysql2/blob/ab3feb75bb4af1f6d288c0494a99aa2483ac24d5/lib/connection_config.js#L171)

**Comment by thibauds** [comment]
@sidorares Thank you for your answer. It is weird because I am getting the following error:
`Can't create more than max_prepared_stmt_count statements (current value: 16382)`
Maybe the cache eviction mechanism cannot keep up with the speed of insertions?

**Comment by thibauds** [comment]
Ok I think I just got it, sorry for the misunderstanding.
The LRU mechanism is on the client side I thought it was on the server side.
Lowering the `max_prepared_stmt_count` seems to the answer :)

**Comment by sidorares** [comment]
> Maybe the cache eviction mechanism cannot keep up with the speed of insertions?

`statement.close()` unfortunately does not have success/error confirmation at the protocol level, so the client expects the PS to be unprepared as soon as "close statement" command is sent, so I can potentially see some out of sync situation here

Yes, number of clients times `maxPreparedStatements` in each client config should give you a possible maximum number of prepared statements clients can attempt to create simultaneously

**Comment by wAkExGOD** [comment]
> Yes, number of clients times `maxPreparedStatements` in each client config should give you a possible maximum number of prepared statements clients can attempt to create simultaneously

Hi! Sorry, but I didn't understand correctly. What should I do to solve the problem `Can't create more than max_prepared_stmt_count statements (current value: 16382)`? For example, I have created the pool with `connectionLimit: 10` and `max_prepared_stmt_count` is default (16382). Should I just lower `max_prepared_stmt_count` in MySQL config, how will it help?

**Comment by sidorares** [comment]
@wAkExGOD need to start with understanding what is the reason you have so many prepared statements. Is it because you have lots of open connections or lots of different versions of similar sql query that can't be parametrised?

**Comment by wAkExGOD** [comment]
@sidorares , Maybe it is the first reason.

My connection to db:
```
import * as mysql from 'mysql2/promise';

const connection = mysql.createPool({
  connectionLimit: 10,
  host: process.env.DB_HOST,
  user: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  database: 'name_of_db'
});
```

Then I make database queries in 3 ways:
1) `const [users] = await connection.execute(" SELECT * FROM users WHERE email = ?", [email] );`
2) `await connection.execute( "UPDATE users SET isActivated = true WHERE userId = ?", [user.userId] );`
3) `await connection.query( "DELETE FROM users WHERE userId = ?", [userId], (error) => { ... } );`


I think, I don't do something wrong. But I can see, that on every page refresh `Prepared_stmt_count` (**show global status like '%prepared%'**) in DB is growing. I thought that I should close connections. But `pool.query / pool.execute` are the same as `pool.getConnection() -> connection.query() -> connection.release()`.

**Comment by sidorares** [comment]
In the above example I'd expect at most 2 PS created per connection ( `SELECT * FROM users WHERE email = ?` and `UPDATE users SET isActivated = true WHERE userId = ?` ). The total number would be 2 * number of open connections in each pool * number of running scripts
